Kuadrant/kuadrant-operator
# RLP atomic override

# Description
Closes: https://github.com/Kuadrant/kuadrant-operator/issues/463

RLP atomic override. RLP targetting gateways that define an override block, will override the limits of HTTPRoute RLPs.

CEL validation was added to:
* make `defaults` and `override` block mutually exclusive
* make implicit limits block and `override` block mutually exclusive
* only allow `override` block for RLPs targetting Gateway resource

# Verification 
The scenario's is already generally tested with the integration tests added 

If you want to manually verify:
* Checkout this branch and deploy cluster
```sh
make local-setup
```
* Deploy Kuadrant CR
```
kubectl -n kuadrant-system apply -f - <<EOF                 
apiVersion: kuadrant.io/v1beta1                       
kind: Kuadrant
metadata:
  name: kuadrant
spec: {}
EOF
```

* Deploy toystore
```sh
kubectl apply -f examples/toystore/toystore.yaml
kubectl wait --timeout=300s --for=condition=Available deployment toystore
```

* Create  HTTP Route
```sh
kubectl apply -f - <<EOF
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: toystore
spec:
  parentRefs:
  - name: istio-ingressgateway
    namespace: istio-system
  hostnames:
  - api.toystore.com
  rules:
  - matches:
    - method: GET
      path:
        type: PathPrefix
        value: "/toys"
    backendRefs:
    - name: toystore
      port: 80
  - matches: # it has to be a separate HTTPRouteRule so we do not rate limit other endpoints
    - method: POST
      path:
        type: Exact
        value: "/toys"
    backendRefs:
    - name: toystore
      port: 80
EOF
```

* Create GW RLP with defaults
```sh
kubectl apply -f - <<EOF
apiVersion: kuadrant.io/v1beta2
kind: RateLimitPolicy
metadata:
  name: gw-rlp
  namespace: istio-system
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: Gateway
    name: istio-ingressgateway
  defaults:
    limits:
      "gateway":
        rates:
        - limit: 5
          duration: 10
          unit: second
EOF
```
* Port forward gateway service
```sh
kubectl port-forward -n istio-system service/istio-ingressgateway-istio 9080:80 2>&1 >/dev/null &
export GATEWAY_URL=localhost:9080
```
* Ensure HTTP Route is rate limit at 5 requests per 10 seconds
```sh
while :; do curl --write-out '%{http_code}\n' --silent --output /dev/null -H 'Host: api.toystore.com' http://$GATEWAY_URL/toys -X POST | grep -E --color "\b(429)\b|$"; sleep 1; done
```
* Create HTTPRoute RLP with implicit default limits
```sh
kubectl apply -f - <<EOF
apiVersion: kuadrant.io/v1beta2
kind: RateLimitPolicy
metadata:
  name: toystore
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: toystore
  limits:
    "route":
      rates:
      - limit: 8
        duration: 10
        unit: second
EOF
```
* Ensure HTTP Route is rate limted at 8 requests per second (Route RLP defaults takes precedence over gateway defaults) (Note: you might need to port forward the gateway service again)
```sh
while :; do curl --write-out '%{http_code}\n' --silent --output /dev/null -H 'Host: api.toystore.com' http://$GATEWAY_URL/toys -X POST | grep -E --color "\b(429)\b|$"; sleep 1; done
```
* Update GW RLP to set overrides instead
```sh
kubectl apply -f - <<EOF
apiVersion: kuadrant.io/v1beta2
kind: RateLimitPolicy
metadata:
  name: gw-rlp
  namespace: istio-system
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: Gateway
    name: istio-ingressgateway
  overrides:
    limits:
      "gateway":
        rates:
        - limit: 5
          duration: 10
          unit: second
EOF
```
* Ensure HTTP Route is rate limits at 5 requests per second again (GW RLP Override takes precedence)
```sh
while :; do curl --write-out '%{http_code}\n' --silent --output /dev/null -H 'Host: api.toystore.com' http://$GATEWAY_URL/toys -X POST | grep -E --color "\b(429)\b|$"; sleep 1; done
```
* Verify overrides to not allowed on RLPs targetting HTTPRoutes
```sh
kubectl apply -f - <<EOF
apiVersion: kuadrant.io/v1beta2
kind: RateLimitPolicy
metadata:
  name: toystore
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: toystore
  overrides:
    limits:
      "create-toy":
        rates:
        - limit: 8
          duration: 10
          unit: second
EOF
```
* Verify explicit defaults and overrides blocks are mutually exclusive
```
kubectl apply -f - <<EOF
apiVersion: kuadrant.io/v1beta2
kind: RateLimitPolicy
metadata:
  name: gw-rlp
  namespace: istio-system
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: Gateway
    name: istio-ingressgateway
  defaults:
    limits:
      "create-toy":
        rates:
        - limit: 1
          duration: 10
          unit: second
  overrides:
    limits:
      "create-toy":
        rates:
        - limit: 5
          duration: 10
          unit: second
EOF
```
* Verify implicit defaults and overrides blocks are mutually exclusive
```sh
kubectl apply -f - <<EOF
apiVersion: kuadrant.io/v1beta2
kind: RateLimitPolicy
metadata:
  name: gw-rlp
  namespace: istio-system
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: Gateway
    name: istio-ingressgateway
  limits:
    "create-toy":
      rates:
      - limit: 1
        duration: 10
        unit: second
  overrides:
    limits:
      "create-toy":
        rates:
        - limit: 5
          duration: 10
          unit: second
EOF
```
